core/src/test/java/hudson/util/AtomicFileWriterTest.java

```diff
@@ -124,4 +140,55 @@ public void badPath() throws Exception {
                        containsString("exists and is neither a directory nor a symlink to a directory"));
         }
     }
+
+    @Issue("JENKINS-48407")
+    @Test
+    public void checkPermissions() throws IOException, InterruptedException {
+
+        final File newFile = tmp.newFile();
+
+        // Check Posix calls are supported (to avoid running this test on Windows for instance)
+        boolean posixSupported = true;
+        try {
+            Files.getPosixFilePermissions(newFile.toPath());
+        } catch (UnsupportedOperationException e) {
+            posixSupported = false;
+        }
+        assumeThat(posixSupported, is(true));
+
+        // given
+        final Set<PosixFilePermission> givenPermissions = EnumSet.of(OWNER_READ,
+                                                                     OWNER_WRITE,
+                                                                     GROUP_READ,
+                                                                     GROUP_WRITE,
+                                                                     OTHERS_READ
+        );
+
+        final Set<PosixFilePermission> notGivenPermissions = EnumSet.of(OWNER_EXECUTE,
+                                                                        GROUP_EXECUTE,
+                                                                        OTHERS_WRITE,
+                                                                        OTHERS_EXECUTE);
+
+        Files.setPosixFilePermissions(newFile.toPath(), givenPermissions);
+        Path filePath = newFile.toPath();
+
+        // when
+        AtomicFileWriter w = new AtomicFileWriter(filePath, StandardCharsets.UTF_8);
+        w.write("whatever");
+        w.commit();
+
+        // then
+        assertFalse(w.getTemporaryPath().toFile().exists());
+        assertTrue(filePath.toFile().exists());
+
+        final Set<PosixFilePermission> posixFilePermissions = Files.getPosixFilePermissions(filePath);
+
+        for (PosixFilePermission perm : givenPermissions) {
+            assertTrue("missing: " + perm, posixFilePermissions.contains(perm));
```

Comment:
But if you don't use a umams of 022 won't this fail?

Reply:
Umask

Reply:
As I understand it, `umask` only applies to file creation. If you change permissions after creation, as done here with `setPermissions`, then `umask` is unused and your permissions are set exactly.

Reply:
I strongly suspect James is right if the pre nio tmp files are created with typical permissions inherited from umask. I.e. that test assumes tmp files will always be world readable when actually this is likely to depend on umask values before running this test. 
Will check/fix later if this surfaces (otp now). Nothing critical as the production code does anyway what was done previously. It may just be that that test is flaky crap.

Reply:
The pre NIO code sets 666 on files, and umask is applied by whatever native code creates files.

It looks like James was right, `givenPermissions` are set exactly on the file before the write, but AtomicFileWriter creates a new file, so the actual permissions only match `givenPermissions` if the umask is 002. I think to make it not depend on umask, we could create a file with `666` (i.e. create temp file with old IO), then read the permissions from disk to get them post-umask, and use those for `givenPermissions`.